include/parlay/type_traits.h

```diff
@@ -230,56 +246,183 @@ inline constexpr bool is_trivial_allocator_v = is_trivial_allocator<Alloc, T>::v
+    ranges of elements and individual elements.
 
-    See proposal D1144R0 for copious details:
-    https://quuxplusone.github.io/blog/code/object-relocation-in-terms-of-move-plus-destroy-draft-7.html
+    See proposal P1144 for copious details:
+    https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p1144r9.html
 */
 
-template <typename T>
-struct is_trivially_relocatable :
-        std::bool_constant<std::is_trivially_move_constructible<T>::value &&
-                           std::is_trivially_destructible<T>::value> { };
+// Standard language support for trivially relocatable (currently P1144)
+#if defined(__cpp_lib_trivially_relocatable)
 
-template <typename T> struct is_nothrow_relocatable :
-        std::bool_constant<is_trivially_relocatable<T>::value ||
-                           (std::is_nothrow_move_constructible<T>::value &&
-                            std::is_nothrow_destructible<T>::value)> { };
+template<typename T>
+using is_trivially_relocatable = std::is_trivially_relocatable<T>;
+
+// Clang's builtin with some (very limited) support for trivial relocatability
+#elif defined(__has_builtin)
+#if __has_builtin(__is_trivially_relocatable)
+
+template<typename T>
+struct is_trivially_relocatable : std::bool_constant<__is_trivially_relocatable(T)> { };
+
+#define PARLAY_MUST_SPECIALIZE_IS_TRIVIALLY_RELOCATABLE
+
+#else
+
+template<typename T>
+struct is_trivially_relocatable : std::is_trivially_copyable<T> { };
+
+#define PARLAY_MUST_SPECIALIZE_IS_TRIVIALLY_RELOCATABLE
+
+#endif
+#else
+
+template<typename T>
+struct is_trivially_relocatable : std::is_trivially_copyable<T> { };
+
+#define PARLAY_MUST_SPECIALIZE_IS_TRIVIALLY_RELOCATABLE
+
+#endif
 
 template<typename T>
 inline constexpr bool is_trivially_relocatable_v = is_trivially_relocatable<T>::value;
 
+#if defined(__cpp_impl_trivially_relocatable)
+#define PARLAY_TRIVIALLY_RELOCATABLE [[trivially_relocatable]]
+#else
+#define PARLAY_TRIVIALLY_RELOCATABLE
+#endif
+
+
+#if defined(PARLAY_MUST_SPECIALIZE_IS_TRIVIALLY_RELOCATABLE)
+
+#define PARLAY_ASSUME_TRIVIALLY_RELOCATABLE(...)                              \
+  struct is_trivially_relocatable<__VA_ARGS__> : std::true_type {};
+
+#define PARLAY_ASSUME_TRIVIALLY_RELOCATABLE_IF(cond, ...)                     \
+  struct is_trivially_relocatable<__VA_ARGS__> : std::bool_constant<cond> {};
+
+
+// We list all the types that ought to be trivially relocatable under any
+// sensible implementation. It is possible that some of these are wrong.
+
+template<typename T>
+PARLAY_ASSUME_TRIVIALLY_RELOCATABLE(std::allocator<T>);
+
+template<typename T>
+PARLAY_ASSUME_TRIVIALLY_RELOCATABLE(std::shared_ptr<T>);
+
+template<typename T>
+PARLAY_ASSUME_TRIVIALLY_RELOCATABLE(std::weak_ptr<T>);
+
+template<typename T>
+PARLAY_ASSUME_TRIVIALLY_RELOCATABLE(std::optional<T>);
+
+template<typename T, typename Deleter>
+PARLAY_ASSUME_TRIVIALLY_RELOCATABLE_IF((is_trivially_relocatable_v<Deleter>), std::unique_ptr<T, Deleter>);
```

Comment:
Technically you also need to check `is_trivially_relocatable_v<Deleter::pointer>`; or just drop the `Deleter` arg the way you dropped the allocator from all the STL containers below.